packages/dds/tree/src/test/shared-tree/sharedTree.spec.ts

```diff
@@ -310,6 +317,77 @@ describe("SharedTree", () => {
             }
         });
 
+        it("can abandon a transaction", async () => {
+            const provider = await TestTreeProvider.create(2);
+            const [tree1] = provider.trees;
+
+            const initialState: JsonableTree = {
+                type: brand("Node"),
+                fields: {
+                    foo: [
+                        { type: brand("Number"), value: 0 },
+                        { type: brand("Number"), value: 1 },
+                        { type: brand("Number"), value: 2 },
+                    ],
+                },
+            };
+            initializeTestTree(tree1, initialState);
+            tree1.runTransaction((forest, editor) => {
+                const rootField = editor.sequenceField(undefined, rootFieldKeySymbol);
+                const root0Path = {
+                    parent: undefined,
+                    parentField: rootFieldKeySymbol,
+                    parentIndex: 0,
+                };
+                const root1Path = {
+                    parent: undefined,
+                    parentField: rootFieldKeySymbol,
+                    parentIndex: 1,
+                };
+                const foo0 = editor.sequenceField(root0Path, fooKey);
+                const foo1 = editor.sequenceField(root1Path, fooKey);
+                editor.setValue(
+                    {
+                        parent: root0Path,
+                        parentField: fooKey,
+                        parentIndex: 1,
+                    },
+                    41,
+                );
+                editor.setValue(
+                    {
+                        parent: root0Path,
+                        parentField: fooKey,
+                        parentIndex: 2,
+                    },
+                    42,
+                );
+                editor.setValue(root0Path, "RootValue1");
+                foo0.delete(0, 1);
+                rootField.insert(0, singleTextCursor({ type: brand("Test") }));
+                foo1.delete(0, 1);
+                editor.setValue(root1Path, "RootValue2");
+                foo1.insert(0, singleTextCursor({ type: brand("Test") }));
+                editor.setValue(
+                    {
+                        parent: root1Path,
+                        parentField: fooKey,
+                        parentIndex: 1,
+                    },
+                    82,
+                );
+                // Aborting the transaction should restore the forest
+                return TransactionResult.Abort;
+            });
+
+            const readCursor = tree1.forest.allocateCursor();
+            moveToDetachedField(tree1.forest, readCursor);
+            assert(readCursor.firstNode());
+            const actual = jsonableTreeFromCursor(readCursor);
+            assert.deepEqual(actual, initialState);
+            readCursor.free();
```

Comment:
It would be nice if we generalized `validateTree` to take the expected tree instead of just a value array. Then we could just call `validateTree(tree1, initialState)`.